pennylane/_qubit_device.py

```diff
@@ -0,0 +1,207 @@
+
+                elif obs.return_type is Sample:
+                    results.append(np.array(self.sample(obs)))
+
+                elif obs.return_type is Probability:
+                    results.append(list(self.probability(wires=obs.wires).values()))
+
+                elif obs.return_type is not None:
+                    raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+
+            self.post_measure()
+
+            self._op_queue = None
+            self._obs_queue = None
+            self._parameters = None
+
+            # Ensures that a combination with sample does not put
+            # expvals and vars in superfluous arrays
+            if all(obs.return_type is Sample for obs in observables):
+                return self._asarray(results)
+            if any(obs.return_type is Sample for obs in observables):
+                return self._asarray(results, dtype="object")
+
+            return self._asarray(results)
+
+    def rotate_basis(self, observable):
+        """Rotates the specified wires such that they
+        are in the eigenbasis of the provided observable.
+
+        Args:
+            observable (Observable): the observable that is to be
+                measured
+        """
+        for diag_gate in observable.diagonalizing_gates()[::-1]:
+            self.apply(diag_gate)
+
+    def expval(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact expectation value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals @ prob).real
+
+        # estimate the ev
+        return np.mean(self.sample(observable))
+
+    def var(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact variance value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals ** 2) @ prob - (eigvals @ prob).real ** 2
+
+        return np.var(self.sample(observable))
+
+    def sample(self, observable):
+        if isinstance(observable, qml.Identity):
+            return np.ones([self.shots])
+        self.rotate_basis(observable)
+        # Sampling is supported on software simulators
+        # HW devices will have to provide their own sample() method
+        if isinstance(observable, qml.Hermitian):
+            eigvals = observable.eigvals(np.array(observable.parameters))
+        else:
+            eigvals = observable.eigvals
+
+        eigvals = np.fromiter(eigvals, dtype=np.float64)
+        prob = np.abs(self._state ** 2)
+        prob = self.marginal_prob(prob, wires=observable.wires)
+        return np.random.choice(eigvals, self.shots, p=prob)
```

Comment:
As far as I can tell, this doesn't fix the inconsistent sampling behaviour, right? One way to fix this is the following:

1. At some point **before** ``sample()`` is called, check if there exists a 'sample' observable in the queue. If so, set a boolean flag, and use the result of this flag to 'pre-generate' all the computational basis samples:
    ```python
    for e in self.obs_queue:
        if hasattr(e, "return_type") and e.return_type == Sample:
            self._memory = True  # make sure to return samples

    # generate computational basis samples
    if self._memory:
        # sample from the computational basis states based on the state probability
        basis_states = np.arange(2**self.num_wires)
        samples = np.random.choice(basis_states, self.shots, p=self.rotated_probability)

        # convert the basis states from base 10 to binary representation
        self._samples = (((samples[:,None] & (1 << np.arange(2**self.num_wires)))) > 0).astype(int)
    ```

2. The `sample()` method would then post-process the resulting computational basis states, that contain correlation information:
    ```python
    def sample(self, observable):
        if observable.name in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
            # observables all have eigenvalues {1, -1}, so post-processing step is known
            return 1 - 2 * self._samples[:, wires[0]]

        # Need to post-process the samples using the observables.
        # Extract only the columns of the basis samples required based on `wires`.
        wires = np.hstack(wires)
        samples = self._samples[:, np.array(wires)]
       
        # replace the basis state in the computational basis with the correct eigenvalue
        indices = np.ravel_multi_index(samples.T, [2] * len(wires))
        converted_measurements = observable.eigvals[indices]
        return converted_measurements
    ```

Reply:
The code in (1) should only run on non-hardware devices though.

The code in 2 should generalize to all devices (hardware and software), as long as self._samples stores the full computational basis samples after execution.

Reply:
As per our discussion, this will be TODO for the sampling PR too